contentprovider.go

```diff
@@ -316,19 +318,24 @@ func (p *contentProvider) fillContentChunkMatches(ms []*candidateMatch, numConte
 		for _, cm := range chunk.candidates {
 			startOffset := cm.byteOffset
 			endOffset := cm.byteOffset + cm.byteMatchSz
-			startLine, startLineOffset, _ := newlines.atOffset(startOffset)
-			endLine, endLineOffset, _ := newlines.atOffset(endOffset)
+			startLine := newlines.atOffset(startOffset)
+			endLine := newlines.atOffset(
+				// We want the line of the last byte in the match, not the first byte outside of the match.
+				// For a zero-length match, endOffset-1 could be before match start, so fall back to the
+				// byte after the match (as we do for startLine), not before.
+				max(startOffset, max(endOffset, 1)-1 /* prevent underflow */),
```

Comment:
you do this in one other place where you need to protect for underflow. Maybe it makes sense for a little helper like
```
startLine, endLine := newlines.atOffsetInclusiveRange(startOffset, endOffset)
```